src/librustc/traits/trans/mod.rs

```diff
@@ -0,0 +1,73 @@
+use dep_graph::{DepGraph, DepNode, DepTrackingMap, DepTrackingMapConfig};
+use hir::def_id::DefId;
+use std::cell::RefCell;
+use std::marker::PhantomData;
+use traits::Vtable;
+use ty::{self, Ty};
+
+/// Specializes caches used in trans -- in particular, they assume all
+/// types are fully monomorphized and that free regions can be erased.
```

Comment:
The monomorphic condition was fixed a while back I think, at least in terms of specialization - and I assume the regions wouldn't matter after we get lazy normalization? Still, can't these caches be chosen dynamically, if all conditions are met?